zocalo/service/dispatcher.py

```diff
@@ -0,0 +1,250 @@
+            recipes = []
+            if message.get("custom_recipe"):
+                try:
+                    recipes.append(
+                        workflows.recipe.Recipe(
+                            recipe=json.dumps(message["custom_recipe"])
+                        )
+                    )
+                except Exception as e:
+                    self.log.error(
+                        "Rejected message containing a custom recipe that caused parsing errors: %s",
+                        str(e),
+                        exc_info=True,
+                    )
+                    self._transport.nack(header)
+                    return
+            if message.get("recipes"):
+                for recipefile in message["recipes"]:
+                    try:
+                        with open(
+                            os.path.join(self._recipe_basepath, recipefile + ".json"),
+                            "r",
+                        ) as rcp:
+                            recipes.append(workflows.recipe.Recipe(recipe=rcp.read()))
+                    except ValueError as e:
+                        self.log.error(
+                            "Error reading recipe '%s': %s", recipefile, str(e)
+                        )
+                        self._transport.nack(header)
+                        return
+                    except IOError as e:
+                        if e.errno == errno.ENOENT:
+                            self.log.error(
+                                "Message references non-existing recipe '%s'",
+                                recipefile,
+                            )
+                            self._transport.nack(header)
+                            return
+                        raise
+
+            if not recipes:
+                self.log.error(
+                    "Message contains no valid recipes or pointers to recipes"
+                )
+                self._transport.nack(header)
+                return
+
+            full_recipe = workflows.recipe.Recipe()
+            for recipe in recipes:
+                try:
+                    recipe.validate()
+                except workflows.Error as e:
+                    self.log.error(
+                        "Recipe failed validation. %s", str(e), exc_info=True
+                    )
+                    self._transport.nack(header)
+                    return
+                recipe.apply_parameters(parameters)
+                full_recipe = full_recipe.merge(recipe)
+
+            # Conditionally acknowledge receipt of the message
+            txn = self._transport.transaction_begin()
+            self._transport.ack(header, transaction=txn)
+
+            rw = workflows.recipe.RecipeWrapper(
+                recipe=full_recipe, transport=self._transport
+            )
+            rw.environment = {
+                "ID": recipe_id
+            }  # FIXME: This should go into the constructor, but workflows can't do that yet
```

Comment:
I've just made that change, so can replace the 6 lines with
```python
rw = workflows.recipe.RecipeWrapper(
                environment={"ID": recipe_id},
                recipe=full_recipe,
                transport=self._transport,
            )
```

Reply:
Done